WhyNotEarth.Meredith.App/Controllers/Api/v0/CompanyController.cs

```
@@ -1,24 +1,47 @@
 namespace WhyNotEarth.Meredith.App.Controllers.Api.v0
 {
+    using System;
     using System.Linq;
     using System.Threading.Tasks;
     using Microsoft.AspNetCore.Cors;
     using Microsoft.AspNetCore.Mvc;
     using Microsoft.EntityFrameworkCore;
+    using WhyNotEarth.Meredith.App.Models.Api.v0.Company;
     using WhyNotEarth.Meredith.Data.Entity;
+    using WhyNotEarth.Meredith.Public;
 
     [ApiVersion("0")]
     [Route("/api/v0/companies")]
     [EnableCors]
     public class CompanyController : Controller
     {
         private MeredithDbContext MeredithDbContext { get; }
+        private CompanyService CompanyService { get; }
 
-        public CompanyController(MeredithDbContext meredithDbContext)
+        public CompanyController(MeredithDbContext meredithDbContext, CompanyService publicService)
         {
+            CompanyService = publicService;
             MeredithDbContext = meredithDbContext;
         }
 
+        [Route("/")]
+        [HttpPost]
+        public async Task<IActionResult> Create(CompanyModel company)
+        {
+            if (!ModelState.IsValid)
+                return BadRequest();
+
+            try
+            {
+                var newCompany = await CompanyService.CreateCompanyAsync(company.Name, company.Slug);
+                return Ok(new { CompanyId = newCompany.Id });
+            }
+            catch (Exception e)
+            {
+                return StatusCode(500, new { error = e.Message });
```

Arguably if we're going to do generic 500s we'll do that in middleware, though with error  handling we should never just catch/display all exceptions: https://www.owasp.org/index.php/Error_Handling, hence why I've created a few internal exceptions that we can assume are "safe" to expose.

---

Ok that makes sense. Should I replace all the ArgumentExceptions/ArgumentNullExceptions I threw with the InvalidActionException, and then here in the controller catch only the InvalidActionExceptions?

---

Depends -- `ArgumentNullExceptions` should be prevented via a properly defined model, and we'll still see `ArgumentNullExceptions` logged in Rollbar (I can get you access to that later).

Basically: if it's something the user/end application can maybe handle, and isn't something that should be caught by a "higher" level of the app (eg: `ArgumentNullExceptions` should _never_ throw because the MVC models should prevent that), throw an `InvalidActionException` (eg: something is in a bad state), if it's internal, feel free to use the normal exceptions and we'll catch/log those as needed.

I'm also not a fan of `InvalidActionException` and I may poke someone on my team that isn't nearly as bad at naming things as myself.